### PR TITLE
Feature create interfaces

### PR DESCRIPTION
**Pull request by ybs007**

### CODE REVIEW

File: Blizzard.IO.Core/IConsumer.cs [hunk 1]
```diff
@@ -0,0 +1,11 @@
+﻿using System;
+
+namespace Blizzard.IO.Core
+{
+    public interface IConsumer<TData>
+    {
+        void StartConsume();
+        void StopConsume();
+        event Action<TData> MessageConsumed;
```
Comment (alloncm) [inline Blizzard.IO.Core/IConsumer.cs]: We forgot to update the design but they requested to call it Start and Stop and to call the event MessageReceived

File: Blizzard.IO.Core/IConsumerWithMetadata.cs [hunk 2]
```diff
@@ -0,0 +1,9 @@
+﻿using System;
+
+namespace Blizzard.IO.Core
+{
+    public interface IConsumerWithMetadata<TData, TMetadata> where TMetadata : BaseMetadata
+    {
+        event Action<TData, TMetadata> MessageWithMetadataConsumed;
```
Comment (alloncm) [inline Blizzard.IO.Core/IConsumerWithMetadata.cs]: Like I said earlier change to revceived

File: Blizzard.IO.Core/IDeserializer.cs [hunk 3]
```diff
@@ -0,0 +1,7 @@
+﻿namespace Blizzard.IO.Core
+{
+    public interface IDeserializer<TData>
+    {
+        TData Desirialize(byte[] data);
```
Comment (alloncm) [inline Blizzard.IO.Core/IDeserializer.cs]: you have a type it is Deserialize
```suggestion
        TData Deserialize(byte[] data);
```

File: Blizzard.IO.Core/ITypeDeserializer.cs [hunk 4]
```diff
@@ -0,0 +1,9 @@
+﻿using System;
+
+namespace Blizzard.IO.Core
+{
+    public interface ITypeDeserializer<TData>
```
Comment (alloncm) [inline Blizzard.IO.Core/ITypeDeserializer.cs]: Myabe Call it ConcreteTypeDeserializer?
```suggestion
    public interface IConcreteTypeDeserializer<TData>
```

File: Blizzard.IO.Core/Rpc/IRpcClient.cs [hunk 5]
```diff
@@ -0,0 +1,10 @@
+﻿using System.Threading.Tasks;
+
+namespace Blizzard.IO.Core.Rpc
+{
+    public interface IRpcClient
+    {
+        TRes Request<TReq, TRes>(TReq request);
+        Task<TRes> RequestASync<TReq, TRes>(TReq request);
```
Comment (alloncm) [inline Blizzard.IO.Core/Rpc/IRpcClient.cs]: Maybe change the TRes and TReq to TRequest and TRespond?
```suggestion
        TRespond Request<TRequest , TRespond>(TRequest request);
        Task<TRespond> RequestASync<TRequest , TRespond>(TRequest request);
```

File: Blizzard.IO.Core/Rpc/IRpcServer.cs [hunk 6]
```diff
@@ -0,0 +1,11 @@
+﻿using System;
+using System.Threading.Tasks;
+
+namespace Blizzard.IO.Core.Rpc
+{
+    public interface IRpcServer
+    {
+        void Respond<TReq, TRes>(Func<TReq, TRes> callback);
+        void RespondAsync<TReq, TRes>(Func<TReq, Task<TRes>> callback);
```
Comment (alloncm) [inline Blizzard.IO.Core/Rpc/IRpcServer.cs]: Like I suggested in IRpcClient

File: Blizzard.IO.Core/IConsumer.cs (hunk 1 above)
Comment (IdoRozval) [inline Blizzard.IO.Core/IConsumer.cs]: I believe it should be `MessageReceived`

File: Blizzard.IO.Core/IConsumerWithMetadata.cs (hunk 2 above)
Comment (IdoRozval) [inline Blizzard.IO.Core/IConsumerWithMetadata.cs]: I believe it should be `MessageReceived`

File: Blizzard.IO.Core/IPublisher.cs [hunk 7]
```diff
@@ -0,0 +1,7 @@
+﻿namespace Blizzard.IO.Core
+{
+    public interface IPublisher<TData>
```
Comment (IdoRozval) [inline Blizzard.IO.Core/IPublisher.cs]: Add the variant `in` notation on `TData`

File: Blizzard.IO.Core/IConsumer.cs [hunk 8]
```diff
@@ -0,0 +1,11 @@
+﻿using System;
+
+namespace Blizzard.IO.Core
+{
+    public interface IConsumer<TData>
```
Comment (IdoRozval) [inline Blizzard.IO.Core/IConsumer.cs]: add the contravariant `out` notation on `TData`

File: Blizzard.IO.Core/IConsumer.cs (hunk 1 above)
Comment (ybs007) [reply Blizzard.IO.Core/IConsumer.cs]: So the class name is should be Receiver or MessageReceiver?

File: Blizzard.IO.Core/IConsumerWithMetadata.cs (hunk 2 above)
Comment (ybs007) [reply Blizzard.IO.Core/IConsumerWithMetadata.cs]: Same as above

**Review by IdoRozval — COMMENTED** [review]
Reviewed by me

File: Blizzard.IO.Core/IConsumer.cs (hunk 1 above)
Comment (ybs007) [reply Blizzard.IO.Core/IConsumer.cs]: f

File: Blizzard.IO.Core/IConsumerWithMetadata.cs (hunk 2 above)
Comment (ybs007) [reply Blizzard.IO.Core/IConsumerWithMetadata.cs]: f

File: Blizzard.IO.Core/IDeserializer.cs (hunk 3 above)
Comment (ybs007) [reply Blizzard.IO.Core/IDeserializer.cs]: f

File: Blizzard.IO.Core/Rpc/IRpcServer.cs (hunk 6 above)
Comment (ybs007) [reply Blizzard.IO.Core/Rpc/IRpcServer.cs]: f

File: Blizzard.IO.Core/IPublisher.cs [hunk 9]
```diff
@@ -0,0 +1,7 @@
+﻿namespace Blizzard.IO.Core
+{
+    public interface IPublisher<in TData>
```
Comment (alloncm) [inline Blizzard.IO.Core/IPublisher.cs]: Why the in?

File: Blizzard.IO.Core/Rpc/IRpcServer.cs (hunk 6 above)
Comment (alloncm) [reply Blizzard.IO.Core/Rpc/IRpcServer.cs]: You havent fixed it

File: Blizzard.IO.Core/IConsumer.cs [hunk 10]
```diff
@@ -0,0 +1,12 @@
+﻿using System;
+
+namespace Blizzard.IO.Core
+{
+    public interface IConsumer<out TData>
```
Comment (alloncm) [inline Blizzard.IO.Core/IConsumer.cs]: Why the out

File: Blizzard.IO.Core/IPublisher.cs (hunk 9 above)
Comment (ybs007) [reply Blizzard.IO.Core/IPublisher.cs]: Rozy told me to use it so you could send data in the inheritance tree

File: Blizzard.IO.Core/IConsumer.cs (hunk 10 above)
Comment (ybs007) [reply Blizzard.IO.Core/IConsumer.cs]: same as above

File: Blizzard.IO.Core/Rpc/IRpcServer.cs (hunk 6 above)
Comment (ybs007) [reply Blizzard.IO.Core/Rpc/IRpcServer.cs]: I did, maybe try to pull or look at the changes here

File: Blizzard.IO.Core/IPublisher.cs (hunk 9 above)
Comment (alloncm) [reply Blizzard.IO.Core/IPublisher.cs]: Rozi and I talked about it and we did not understood this keyword correctly
we decided not to use it for now

File: Blizzard.IO.Core/IConsumer.cs (hunk 10 above)
Comment (alloncm) [reply Blizzard.IO.Core/IConsumer.cs]: same as above

Comment (ybs007) [reply Blizzard.IO.Core/IConsumer.cs]: f

File: Blizzard.IO.Core/IPublisher.cs (hunk 9 above)
Comment (ybs007) [reply Blizzard.IO.Core/IPublisher.cs]: f

**Pull request merged**